.github/workflows/build.yaml

```diff
@@ -1,26 +1,366 @@
 name: "Test"
 
 on:
-    pull_request:
+  pull_request:
+  push:
+    branches:
+      - master
+    tags:
+      - "v*"
 
 jobs:
-    build:
-        name: "MacOS build"
+    tarball:
+      runs-on: ubuntu-latest
+      name: Tarball
+      steps:
+        - name: Clone wake
+          uses: actions/checkout@v3
+          with:
+            fetch-depth: 0
+
+        - name: Install Deps
+          run: sudo apt-get update && sudo apt-get install -y build-essential fuse libfuse-dev libsqlite3-dev libgmp-dev libncurses5-dev pkg-config git g++ gcc libre2-dev python3-sphinx clang-format-12
+
+        - name: Check Formatting
+          run: clang-format-12 --style=file --Werror -n $(./scripts/which_clang_files all)
+
+        - name: Build Tarball
+          run: make tarball
+
+        - name: Run Tests
+          run: make test && make unittest
+
+        - name: Generate Docs
+          run: mkdir www && ./bin/wake --no-workspace --html > www/index.html
+
+        - name: wake to rst
+          run: PATH=$(pwd)/bin:$PATH PYTHONPATH=$(pwd)/scripts python3 scripts/wake2rst.py
+
+        - name: Upload tarball
+          uses: actions/upload-artifact@v3
```

Comment:
What does the upload-artifact action do?

Reply:
uploads the contents of a folder or path to a key-value store that is a new store for each workflow run